st2common/st2common/util/param.py

```diff
@@ -0,0 +1,212 @@
+    cast = get_cast(cast_type=parameter_type)
+    if not cast:
+        return v
+
+    return cast(v)
+
+
+def _create_graph(action_context):
+    '''
+    Creates a generic directed graph for depencency tree and fills it with basic context variables
+    '''
+    G = nx.DiGraph()
+    G.add_node(SYSTEM_KV_PREFIX, value=KeyValueLookup())
+    G.add_node(ACTION_CONTEXT_KV_PREFIX, value=action_context)
+    return G
+
+
+def _process(G, name, value):
+    '''
+    Determines whether parameter is a template or a value. Adds graph nodes and edges accordingly.
+    '''
+    # Jinja defaults to ascii parser in python 2.x unless you set utf-8 support on per module level
+    # Instead we're just assuming every string to be a unicode string
+    if isinstance(value, str):
+        value = to_unicode(value)
+    template_ast = ENV.parse(value)
+    # Dependencies of the node represent jinja variables used in the template
+    # We're connecting nodes with an edge for every depencency to traverse them in the right order
+    # and also make sure that we don't have missing or cyclic dependencies upfront.
+    dependencies = meta.find_undeclared_variables(template_ast)
+    if dependencies:
+        G.add_node(name, template=value)
+        for dependency in dependencies:
+            G.add_edge(dependency, name)
+    else:
+        G.add_node(name, value=value)
+
+
+def _process_defaults(G, schemas):
+    '''
+    Process dependencies for parameters default values in the order schemas are defined.
+    '''
+    for schema in schemas:
+        for name, value in six.iteritems(schema):
+            if name not in G.node or value.get('immutable', False):
+                _process(G, name, value.get('default', None))
+
+
+def _validate(G):
+    '''
+    Validates dependency graph to ensure it has no missing or cyclic dependencies
+    '''
+    for name in G.nodes():
+        if 'value' not in G.node[name] and 'template' not in G.node[name]:
+            msg = 'Dependecy unsatisfied in %s' % name
+            raise ParamException(msg)
+
+    if not nx.is_directed_acyclic_graph(G):
+        msg = 'Cyclic dependecy found'
+        raise ParamException(msg)
+
+
+def _render(node, render_context):
+    '''
+    Render the node depending on its type
+    '''
+    if 'template' in node:
+        return ENV.from_string(node['template']).render(render_context)
+    if 'value' in node:
+        return node['value']
+
+
+def _resolve_dependencies(G):
+    '''
+    Traverse the dependency graph starting from resolved nodes
+    '''
+    context = {}
+    for name in nx.topological_sort(G):
+        node = G.node[name]
+        try:
+            context[name] = _render(node, context)
+        except Exception as e:
+            LOG.debug('Failed to render %s: %s', name, e, exc_info=True)
+            msg = 'Failed to render parameter "%s": %s' % (name, str(e))
+            raise ParamException(msg)
+    return context
+
+
+def _cast_params_from(params, context, schemas):
```

Comment:
Yeah, that's my plan. The only thing I'm not sure about is whether I should do it now or postpone until we have a clear understanding this thing is working and doesn't require a major changes because it suddenly stopped supporting some crucial use case.